include/swift/AST/AccessScopeChecker.h

```diff
@@ -23,47 +23,79 @@
 
 namespace swift {
 
-class AbstractStorageDecl;
-class ExtensionDecl;
+class BoundGenericType;
+class ComponentIdentTypeRepr;
+class NominalType;
 class SourceFile;
-class ValueDecl;
+class TypeAliasType;
 
 class AccessScopeChecker {
   const SourceFile *File;
   bool TreatUsableFromInlineAsPublic;
 
-protected:
-  ASTContext &Context;
   Optional<AccessScope> Scope = AccessScope::getPublic();
 
   AccessScopeChecker(const DeclContext *useDC,
                      bool treatUsableFromInlineAsPublic);
-  bool visitDecl(ValueDecl *VD);
-};
-
-class TypeReprAccessScopeChecker : private ASTWalker, AccessScopeChecker {
-  TypeReprAccessScopeChecker(const DeclContext *useDC,
-                             bool treatUsableFromInlineAsPublic);
-
-  bool walkToTypeReprPre(TypeRepr *TR) override;
-  bool walkToTypeReprPost(TypeRepr *TR) override;
+  bool visitDecl(const ValueDecl *VD);
 
 public:
   static Optional<AccessScope>
   getAccessScope(TypeRepr *TR, const DeclContext *useDC,
                  bool treatUsableFromInlineAsPublic = false);
+  static Optional<AccessScope>
+  getAccessScope(Type T, const DeclContext *useDC,
+                 bool treatUsableFromInlineAsPublic = false);
 };
 
-class TypeAccessScopeChecker : private TypeWalker, AccessScopeChecker {
-  TypeAccessScopeChecker(const DeclContext *useDC,
-                         bool treatUsableFromInlineAsPublic);
+/// Walks a Type to find all NominalTypes, BoundGenericTypes, and
+/// TypeAliasTypes.
+class TypeDeclFinder : public TypeWalker {
+  Action walkToTypePre(Type T) override;
+
+public:
+  virtual Action visitNominalType(NominalType *ty) {
+    return Action::Continue;
+  }
+  virtual Action visitBoundGenericType(BoundGenericType *ty) {
+    return Action::Continue;
+  }
+  virtual Action visitTypeAliasType(TypeAliasType *ty) {
+    return Action::Continue;
+  }
+};
 
-  Action walkToTypePre(Type T);
+/// A TypeDeclFinder for use cases where all types should be treated
+/// equivalently and where generic arguments can be walked to separately from
+/// the generic type.
+class SimpleTypeDeclFinder : public TypeDeclFinder {
+  /// The function to call when a ComponentIdentTypeRepr is seen.
+  llvm::function_ref<Action(const TypeDecl *)> Callback;
```

Comment:
I would feel more confident if this was an std::function since then you don't have to be sure to keep the closure alive somewhere else

Reply:
:-/ `std::function` has a definite cost, though, and doesn't inline away nicely. I'd rather keep this as is.